### PR TITLE
fix: mpool-pending fixes

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
## Summary of changes

Changes introduced in this pull request:

- random mempool fixes that surfaced when I was investigating some performance optimizations last week, see changelog for details,
- some internal test refactoring

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation. All new code adheres to the team's [documentation standards](https://github.com/ChainSafe/forest/wiki/Documentation-practices),
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

### Outside contributions

- [ ] This pull request is based on an issue that a maintainer has accepted (see [Before Opening a Pull Request][4]).
- [ ] I have read and agree to the [CONTRIBUTING][2] document.
- [ ] I have read and agree to the [AI Policy][3] document. I understand that failure to comply with the guidelines will lead to rejection of the pull request.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md
[2]: https://github.com/ChainSafe/forest/blob/main/CONTRIBUTING.md
[3]: https://github.com/ChainSafe/forest/blob/main/AI_POLICY.md
[4]: https://github.com/ChainSafe/forest/blob/main/CONTRIBUTING.md#before-opening-a-pull-request

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved pending message results during chain reorganizations and forked tipsets.
  - Prevented pending results from getting ahead of their corresponding tipset.
  - Messages with missing or invalid cached BLS signatures are now skipped safely instead of causing failures.
  - Improved handling of null-round ancestry gaps and duplicate messages during fork merging.
  - Improved consistency when combining pending messages across competing chain histories.

### CODE REVIEW

File: src/message_pool/msgpool/utils.rs [hunk 1]
```diff
@@ -54,16 +51,20 @@ pub(in crate::message_pool) fn get_gas_perf(gas_reward: &TokenAmount, gas_limit:
...
-    let val = bls_sig_cache
-        .get(&msg.cid())
-        .ok_or_else(|| Error::Other("Could not recover sig".to_owned()))?;
-    let smsg = SignedMessage::new_from_parts(msg, val)?;
-    Ok(smsg)
+) -> Option<SignedMessage> {
+    let msg_cid = msg.cid();
+    let recovered = bls_sig_cache
+        .get(&msg_cid)
+        .and_then(|sig| SignedMessage::new_from_parts(msg, sig).ok());
```
Comment (akaladarshi) [inline src/message_pool/msgpool/utils.rs]: Do you think we can directly use the `new_unchecked` since we are getting the signature from the cache which should already contain the verified signature. So no need to re-verify it.

**Review by akaladarshi — COMMENTED** [review]
One comment otherwise LGTM

File: src/rpc/methods/mpool.rs [hunk 2]
```diff
@@ -116,34 +116,29 @@ impl RpcMethod<1> for MpoolPending {
...
         }
 
-        let mut have_cids = HashSet::new();
-        for item in pending.iter() {
-            have_cids.insert(item.cid());
-        }
+        let mut have_cids: HashSet<_> = pending.iter().map(|m| m.cid()).collect();
 
         loop {
+            // A null round can make `ts.parents()` skip past `mpts.epoch()`, so this may never
```
Comment (akaladarshi) [inline src/rpc/methods/mpool.rs:122]: Should we add the test for this null round case?

File: src/message_pool/msgpool/utils.rs (hunk 1 above)
Comment (LesnyRumcajs) [reply src/message_pool/msgpool/utils.rs]: Yeah, I think it makes sense.

**Review by akaladarshi — APPROVED** [review]
